stdlib/public/SDK/Foundation/DataProtocol.swift

```diff
@@ -0,0 +1,239 @@
+//
+// Copyright (c) 2018 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+
+#if os(macOS) || os(iOS)
+import Darwin
+#elseif os(Linux)
+import Glibc
+#endif
+
+// -- DataProtocol -------------------------------------------------------------
+public protocol DataProtocol : RandomAccessCollection where Element == UInt8, SubSequence : DataProtocol {
+    // -- Core Requirements ----------------------------------------------------
+    // FIXME: Remove in favor of opaque type on `regions`.
+    associatedtype RegionSequence: Sequence where RegionSequence.Element : DataProtocol & ContiguousCollection
+    
+    /// A `Sequence` of `DataProtocol` elements which compose a discontiguous buffer of memory.
+    /// Each region is a contiguous buffer of bytes.
+    ///
+    /// The sum of the lengths of the associated regions must equal `self.count`
+    /// (such that iterating `regions` and iterating `self` produces the same sequence of indices
+    /// in the same number of index advancements).
+    var regions: RegionSequence { get }
+    
+    // -- Searching ------------------------------------------------------------
+    /// Returns the first found range of the given data buffer.
+    /// A default implementation can be given in terms of `self.regions`.
+    func firstRange<D: DataProtocol, R: RangeExpression>(of: D, in: R) -> Range<Index>? where R.Bound == Index
+    
+    /// Returns the last found range of the given data buffer.
+    /// A default implementation can be given in terms of `self.regions`.
+    func lastRange<D: DataProtocol, R: RangeExpression>(of: D, in: R) -> Range<Index>? where R.Bound == Index
+    
+    // -- Copying --------------------------------------------------------------
+    /// Copies `count` bytes from the start of the buffer the destination buffer.
+    /// Default implementation can be given in terms of `copyBytes(to:from:)`.
+    @discardableResult
+    func copyBytes(to: UnsafeMutableRawBufferPointer, count: Int) -> Int
+    @discardableResult
+    func copyBytes<DestinationType>(to: UnsafeMutableBufferPointer<DestinationType>, count: Int) -> Int
+    
+    /// Copies the bytes from the given range to the destination buffer.
+    /// Default implementation can be given in terms of `self.regions`.
+    @discardableResult
+    func copyBytes<R: RangeExpression>(to: UnsafeMutableRawBufferPointer, from: R) -> Int where R.Bound == Index
+    @discardableResult
+    func copyBytes<DestinationType, R: RangeExpression>(to: UnsafeMutableBufferPointer<DestinationType>, from: R) -> Int where R.Bound == Index
+    
+    
+    var _borrowedBytes: (buffer: UnsafeRawBufferPointer, owner: Any)? { get }
```

Comment:
If we end up with this here, we should add a reminder to actually document this and what it does.